gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystem.java

```diff
@@ -994,21 +994,24 @@ public Path makeQualified(Path path) {
         "Path '%s' must be fully qualified.",
         qualifiedPath);
 
+    Path result;
+    String upath = uri.getPath();
+
     // Strip initial '..'s to make root is its own parent.
-    StringBuilder sb = new StringBuilder(uri.getPath());
-    while (sb.indexOf("/../") == 0) {
+    int i = 0;
+    while (upath.startsWith("/../", i)) {
       // Leave a preceding slash, so path is still absolute.
-      sb.delete(0, 3);
+      i += 3;
     }
-
-    String strippedPath = sb.toString();
-
-    // Allow a Path of gs://someBucket to map to gs://someBucket/
-    if (strippedPath.equals("/..") || strippedPath.equals("")) {
-      strippedPath = "/";
+    if (i == upath.length() || upath.substring(i).equals("/..")) {
```

Comment:
I feel that the new code is harder to understand. How much benefit (in terms of perforamnce) are we getting by having this? Looks like 3 micro seconds? Do we have any benchmarking data to justify this change?

Reply:
agree the original code is easier to read. however, as the prefix "/../" check is for hadoop-2 style and it is not a common case, the original code does extra stringbuilder ops for this. my local benchmark shows 3ms for the common use cases. As we have seen deletion slow (for example flink checkpoint deletion can not finish within time for thousands of files deletion), optimize this is necessary.